File: src/main/java/com/netdetpla/ndp/controller/ImageController.java
package com.netdetpla.ndp.controller;

import com.netdetpla.ndp.bean.Image;
import com.netdetpla.ndp.bean.ResponseEnvelope;
import com.netdetpla.ndp.handler.DatabaseHandler;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Properties;

@RestController
public class ImageController {

    // 保存镜像文件
    private static void writeToLocal(String destination, InputStream input)
            throws IOException {
        int index;
        byte[] bytes = new byte[1024];
        FileOutputStream downloadFile = new FileOutputStream(destination);
        while ((index = input.read(bytes)) != -1) {
            downloadFile.write(bytes, 0, index);
            downloadFile.flush();
        }
        downloadFile.close();
        input.close();
    }

    // 上传镜像
    @PostMapping("/image/upload")
    public ResponseEntity<?> uploadFile(
            @RequestParam("file") MultipartFile uploadFile,
            @RequestParam("image-name") String imageName,
            @RequestParam("tag") String tag,
            @RequestParam("test-param") String testParam
    ) {
        String size = "";
        try {
//            saveUploadedFiles(Collections.singletonList(uploadFile));
            writeToLocal("/root/image/" + uploadFile.getOriginalFilename(), uploadFile.getInputStream());
            size = String.format("%.2fMB", (double) uploadFile.getSize() / 1024 / 1024);
        } catch (IOException e) {
            System.out.println(e.toString());
            return new ResponseEntity<>(new ResponseEnvelope<>(
                    HttpStatus.BAD_REQUEST.value(),
                    "Bad Request"
            ), HttpStatus.BAD_REQUEST);
        }
        // TODO 镜像导入、测试任务
        System.out.println("image name: " + imageName);
        System.out.println("tag: " + tag);
        System.out.println("param: " + testParam);
        Date nowTime = new Date();
        SimpleDateFormat time = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        System.out.println(time.format(nowTime));
        DatabaseHandler.execute(
                "insert into image(image_name, tag, `size`, upload_time, file_name) values (?, ?, ?, ?, ?)",
                imageName,
                tag,
                size,
                time.format(nowTime),
                uploadFile.getOriginalFilename()
        );
        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "Image has been uploaded."
        ), HttpStatus.OK);
    }

    @GetMapping("/image")
    public ResponseEntity<?> getImages() throws SQLException {
        List<String> data = new ArrayList<>();
        ResultSet resultSet = DatabaseHandler.executeQuery("select distinct image_name from image");
        while (resultSet.next()) {
            data.add(resultSet.getString(1));
        }
        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK);
    }

    @GetMapping("/image/{image_name}")
    public ResponseEntity<?> getTags(@PathVariable(value = "image_name") String imageName) throws SQLException {
        List<Image> data = new ArrayList<>();
        ResultSet resultSet = DatabaseHandler.executeQuery(
                "select tag, upload_time, `size` from image where image_name = ?",
                imageName
        );
        while (resultSet.next()) {
            data.add(new Image(
                    resultSet.getString(1),
                    resultSet.getString(2),
                    resultSet.getString(3)
            ));
        }
        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK);
    }

    @GetMapping("/image/{image_name}/{tag}")
    public ResponseEntity<?> getImageInfo(
            @PathVariable(value = "image_name") String imageName,
            @PathVariable String tag
    ) throws SQLException {
        //TODO 查询任务
        ResultSet resultSet = DatabaseHandler.executeQuery(
                "select upload_time, size from image where image_name = ? and tag = ?",
                imageName,
                tag
        );
        resultSet.next();
        // TODO 镜像大小
        // TODO 测试数据
        Image data = new Image(
                imageName,
                tag,
                true,
                resultSet.getString(1),
                resultSet.getString(2),
                0.5f,
                1024,
                20
        );
        return new ResponseEntity<>(new ResponseEnvelope<>(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK);
    }
}

File: src/main/kotlin/com/ndp/server/controller/ImageController.kt
package com.ndp.server.controller

import com.ndp.server.bean.ResponseEnvelope
import com.ndp.server.utils.DatabaseHandler
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.text.SimpleDateFormat
import java.util.*

@RestController
class ImageController {

    // 保存镜像文件
    @Throws(IOException::class)
    private fun writeToLocal(destination: String, input: InputStream) {
        var index: Int
        val bytes = ByteArray(1024)
        val downloadFile = FileOutputStream(destination)
        while (input.read(bytes).also { index = it } != -1) {
            downloadFile.write(bytes, 0, index)
            downloadFile.flush()
        }
        downloadFile.close()
        input.close()
    }

    class Upload(val imageName: String, val tag: String, val testParam: String)

    // 上传镜像
    @PostMapping("/image/upload")

    fun uploadFile(
            @RequestParam("file") file: MultipartFile,
            @RequestParam("image-name") imageName: String,
            @RequestParam("tag") tag: String,
            @RequestParam("test-param") testParam: String
    ): ResponseEntity<*>? {
        println(file.originalFilename)
        val size = try {
            writeToLocal("/root/image/" + file.originalFilename, file.inputStream)
            String.format("%.2fMB", file.size.toDouble() / 1024 / 1024)
        } catch (e: IOException) {
            println(e.toString())
            return ResponseEntity(ResponseEnvelope<Any?>(
                    HttpStatus.BAD_REQUEST.value(),
                    "Bad Request"
            ), HttpStatus.BAD_REQUEST)
        }
        // TODO 镜像导入、测试任务
        println("image name: $imageName")
        println("tag: $tag")
        println("param: $testParam")
        val nowTime = Date()
        val time = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        println(time.format(nowTime))
        DatabaseHandler.insertImage(
                imageName,
                tag,
                size,
                time.format(nowTime),
                file.originalFilename ?: ""
        )
        return ResponseEntity(ResponseEnvelope<Any?>(
                HttpStatus.OK.value(),
                "Image has been uploaded."
        ), HttpStatus.OK)
    }



    @GetMapping("/image")
    fun getImages(): ResponseEntity<*>? {
        val data = DatabaseHandler.selectImage()
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/image/{image_name}")
    fun getTags(
            @PathVariable(value = "image_name") imageName: String
    ): ResponseEntity<*>? {
        val data = DatabaseHandler.selectTags(imageName)
        return ResponseEntity(ResponseEnvelope(
                HttpStatus.OK.value(),
                "OK",
                data
        ), HttpStatus.OK)
    }

    @GetMapping("/image/{image_name}/{tag}")
    fun getImageInfo(
            @PathVariable(value = "image_name") imageName: String,
            @PathVariable tag: String
    ): ResponseEntity<*>? {
        // TODO 暂留接口
        throw UnsupportedOperationException()
    }
}